fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -408,17 +408,36 @@ void DLApi::addNetworkThreadCompletionHook(void (*hook)(void*), void *hookParame
 }
 
 // MultiVersionTransaction
-MultiVersionTransaction::MultiVersionTransaction(Reference<MultiVersionDatabase> db) : db(db) {
+MultiVersionTransaction::MultiVersionTransaction(Reference<MultiVersionDatabase> db, UniqueOrderedOptionList<FDBTransactionOptions> defaultOptions) : db(db) {
+	setDefaultOptions(defaultOptions);
 	updateTransaction();
 }
 
-// SOMEDAY: This function is unsafe if it's possible to set Database options that affect subsequently created transactions. There are currently no such options.
+void MultiVersionTransaction::setDefaultOptions(UniqueOrderedOptionList<FDBTransactionOptions> options) {
+	MutexHolder holder(db->dbState->optionLock);
+	std::copy(options.begin(), options.end(), std::back_inserter(persistentOptions));
+}
+
 void MultiVersionTransaction::updateTransaction() {
 	auto currentDb = db->dbState->dbVar->get();
 
 	TransactionInfo newTr;
 	if(currentDb.value) {
 		newTr.transaction = currentDb.value->createTransaction();
+
+		Optional<StringRef> timeout;
+		for (auto option : persistentOptions) {
+			if(option.first == FDBTransactionOptions::TIMEOUT) {
+				timeout = option.second.castTo<StringRef>();
+			}
+			else {
+				newTr.transaction->setOption(option.first, option.second.castTo<StringRef>());
+			}
+		}
+	
+		if(timeout.present()) {
+			newTr.transaction->setOption(FDBTransactionOptions::TIMEOUT, timeout);
```

Comment:
I think it's the case that setting the timeout can immediately end a transaction if the timeout has elapsed. This can cause a problem if, for example, you set a database default at 1 second and then override it to 5 seconds on the transaction. If the transaction has run for 2 seconds and you apply both timeouts, the database default will then cause you to fail.

The special logic here is to make sure we only set the most recent value of the timeout.

Reply:
I can add a comment to this effect.

Reply:
Hm, wait, if `persistentOptions` were a `UniqueOrderedOptionList`, would that solve this? Or is there some reason it can't be one of those?

Reply:
That would do it, and I could be convinced to go that route for persistent options. The reason I didn't was because it also imposes the constraint that persistent options cannot have cumulative effects (like database defaults), and I made the choice to not do that since this workaround was easy enough.

It's not a position I hold strongly, though, so feel free to change my mind.

Reply:
None of the current persistent options have cumulative effects, right? So, if this went in as is, we could change it to use the `UniqueOrderedOptionList` later? Maybe we should keep it, then, as it doesn't restrict our options.

Reply:
Yes, I believe that to be the case.